### DISCUSSION TITLE
add received_at column in fulfillment table

### DISCUSSION THREAD

**Discussion by eric-makro** [medusajs/medusa#3445] [Feature Requests]
Hi Medusa Team, is it possible to add `received_at` in fulfillment object? 

I think usually when we create an order, after the product is shipped, we should be able to track when the product is received.

Currently we can use `fulfillment.shipped_at` to show the date in timeline, maybe if we have `received_at`, it should be able to be used here
<img width="473" alt="image" src="https://user-images.githubusercontent.com/110385557/224286270-4fd0ed9b-8c9a-4cd4-94a9-5b470d5b1894.png">

<img width="424" alt="image" src="https://user-images.githubusercontent.com/110385557/224286458-9b5e417b-b735-49ec-9e4c-96ff25ef99ec.png">

Please kindly consider it, thanks

**Comment by srindom** [comment]
Sounds like a great idea - I would suggest `delivered_at` instead of `received_at`. 

Is there any other data that you would like to capture when a fulfillment is marked as delivered or is the timestamp the only relevant thing?

**Reply by eric-makro** [reply]
Hi @srindom , currently `delivered_at` should be enough for now, thanks so much!

**Comment by srindom** [comment]
My high-level thinking on the design for this would be:

### Data model
```ts
// models/fulfillment.ts

@Column({ type: resolveDbType("timestamptz"), nullable: true })
delivered_at: Date | null
```

### Service layer

**FulfillmentService**
- Add `markDelivered(id: string, time?: Date)` - should update `delivered_at` - if no time provided use now
- Add `markUndelivered(id: string)` - should update `delivered_at` to `null`
- Should validate that preconditions are met (e.g., fulfillment not canceled, time > created_at, shipped_at not null, etc.)
- Both should emit events `fulfillment.delivered`, `fulfillment.undelivered`.

### API

**Mark a fulfillment as delivered**
```
POST /admin/orders/:id/fulfillments/:fulfillment_id/delivery
```
- Should validate relationship between `id` and `fulfillment_id` 
- Should call `fulfillmentService.markDelivered()`

**Mark a fulfillment as undelivered**
```
DELETE /admin/orders/:id/fulfillments/:fulfillment_id/delivery
```
- Should validate relationship between `id` and `fulfillment_id` 
- Should call `fulfillmentService.markUndelivered()`

### Admin

- An **Items Delivered** event should be added to the order timeline.
- Fulfillment card on order details page should have buttons for managing delivery

We would need @ludvig18 to create designs for this before starting development on Admin UI.

### Testing
Unit and integration tests should be added.

---

If anyone is up for opening a PR on this we would happily help land it please ping here if you are interested :)

**Reply by mrkn16ht** [reply]
Hi @srindom, I created PR https://github.com/medusajs/medusa/pull/3454 that includes Data model, Service layer, API, can you please help review, thanks
